### PR TITLE
feat: define SetRel output nullability derivation (#558)

### PR DESCRIPTION
**Pull request by vbarua**

### CODE REVIEW

**Comment by vbarua** [comment]
I've attempted to clarify the language around SetRels and provide additional information around deriving nullability for the output types.

File: site/docs/relations/logical_relations.md [hunk 1]
```diff
@@ -261,14 +261,35 @@ The set operation encompasses several set-level operations that support combinin
...
-| Union All               | Returns all records from each set, allowing duplicates.      |
+The set operation type determines both the records that are emitted and the type of the output record.
+
+| Property                | Description                                                                                                   | Output Shape
+| ----------------------- | ------------------------------------------------------------------------------------------------------------- | ----------------------------- |
+| Minus (Primary)         | Returns all records from the primary input excluding any matching records from secondary inputs.              | The same as the primary input
+| Minus (Multiset)        | Returns all records from the primary input excluding any records that are included in *all* secondary inputs. | The same as the primary input
+| Intersection (Primary)  | Returns all records from the primary input that match at least one record from *any* secondary inputs.        | If a field is nullabe in in the primary input and in any of the secondary inputs, is nullable in the output.
```
Comment (EpsilonPrime) [inline site/docs/relations/logical_relations.md:270]: nullabe -> nullable

File: site/docs/relations/logical_relations.md [hunk 2]
```diff
@@ -261,14 +261,35 @@ The set operation encompasses several set-level operations that support combinin
...
+| Intersection (Multiset) | Returns all records from the primary input that match at least one record from *all* secondary inputs.        | If a field is non-null in any of the inputs, it is non-null in the output.
```
Comment (EpsilonPrime) [inline site/docs/relations/logical_relations.md:271]: non-null -> not nullable?  Or use required language/constant?

File: site/docs/relations/logical_relations.md [hunk 3]
```diff
@@ -261,14 +261,35 @@ The set operation encompasses several set-level operations that support combinin
...
+| Union Distinct          | Returns all the records from each set, removing any rows that are duplicated (within or across sets).         | If a field is nullable in any of the inputs, it is nullable in the output.
+| Union All               | Returns all records from each set, allowing duplicates.                                                       | If a field is nullable in any of the inputs, it is nullable in the output |
```
Comment (EpsilonPrime) [inline site/docs/relations/logical_relations.md:273]: End with a period for consistency?  (and above)

**Review by EpsilonPrime — APPROVED** [review]

File: site/docs/relations/logical_relations.md (hunk 1 above)
Comment (vbarua) [reply site/docs/relations/logical_relations.md]: Fixed

File: site/docs/relations/logical_relations.md (hunk 3 above)
Comment (vbarua) [reply site/docs/relations/logical_relations.md]: Now that you've pointed it out I cannot unsee it. Updated.

File: site/docs/relations/logical_relations.md (hunk 2 above)
Comment (vbarua) [reply site/docs/relations/logical_relations.md]: I like `required` for this here, it works well with the examples as well.

**Review by EpsilonPrime — DISMISSED** [review]

File: site/docs/relations/logical_relations.md [hunk 4]
```diff
@@ -249,7 +249,7 @@ The set operation encompasses several set-level operations that support combinin
 | Inputs               | 2 or more                                                    |
 | Outputs              | 1                                                            |
 | Property Maintenance | Maintains distribution if all inputs have the same ordinal distribution. Orderedness is not maintained. |
-| Direct Output Order  | The field order of the inputs.  All inputs must have identical fields. |
+| Direct Output Order  | The field order of the inputs. All inputs must have identical field *types*, but field nullabilities may vary |
```
Comment (EpsilonPrime) [inline site/docs/relations/logical_relations.md:252]: a trailing period here would be nice too

File: site/docs/relations/logical_relations.md [hunk 5]
```diff
@@ -261,14 +261,35 @@ The set operation encompasses several set-level operations that support combinin
...
-| Union All               | Returns all records from each set, allowing duplicates.      |
+The set operation type determines both the records that are emitted and the type of the output record.
+
+| Property                | Description                                                                                                   | Output Shape
+| ----------------------- | ------------------------------------------------------------------------------------------------------------- | ----------------------------- |
+| Minus (Primary)         | Returns all records from the primary input excluding any matching records from secondary inputs.              | The same as the primary input.
+| Minus (Multiset)        | Returns all records from the primary input excluding any records that are included in *all* secondary inputs. | The same as the primary input.
+| Intersection (Primary)  | Returns all records from the primary input that match at least one record from *any* secondary inputs.        | If a field is nullable in in the primary input and in any of the secondary inputs, is nullable in the output.
```
Comment (westonpace) [inline site/docs/relations/logical_relations.md:270]:
```suggestion
| Intersection (Primary)  | Returns all records from the primary input that match at least one record from *any* secondary inputs.        | If a field is nullable in the primary input and in any of the secondary inputs, it is nullable in the output.
```

File: site/docs/relations/logical_relations.md [hunk 6]
```diff
@@ -261,14 +261,35 @@ The set operation encompasses several set-level operations that support combinin
...
-| ----------------------- | ------------------------------------------------------------ |
-| Minus (Primary)         | Returns the primary input excluding any matching records from secondary inputs. |
-| Minus (Multiset)        | Returns the primary input minus any records that are included in all sets. |
-| Intersection (Primary)  | Returns all rows primary rows that intersect at least one secondary input. |
-| Intersection (Multiset) | Returns all rows that intersect at least one record from each secondary inputs. |
-| Union Distinct          | Returns all the records from each set, removing any rows that are duplicated (within or across sets). |
-| Union All               | Returns all records from each set, allowing duplicates.      |
+The set operation type determines both the records that are emitted and the type of the output record.
+
+| Property                | Description                                                                                                   | Output Shape
```
Comment (westonpace) [inline site/docs/relations/logical_relations.md:266]:
```suggestion
| Property                | Description                                                                                                   | Output Nullability
```
"Shape" is a little vague

File: site/docs/relations/logical_relations.md [hunk 7]
```diff
@@ -261,14 +261,35 @@ The set operation encompasses several set-level operations that support combinin
...
+Input 3: (R, N, R, N, R, N, R, N)  Secondary Input
+```
+
+The output type is as follows for the various operations
+
+| Property                | Output Type
+| ----------------------- | -----------------------------------------------------------------------------------------------------
+| Minus (Primary)         | (R, R, R, R, N, N, N, N)
+| Minus (Multiset)        | (R, R, R, R, N, N, N, N)
+| Intersection (Primary)  | (R, R, R, R, R, N, R, N)
```
Comment (westonpace) [inline site/docs/relations/logical_relations.md:289]:
```suggestion
| Intersection (Primary)  | (R, R, R, R, R, N, N, N)
```

**Review by westonpace — APPROVED** [review]
Looks good (modulo a few minor things).  Thanks for clarifying.

**Comment by westonpace** [comment]
Does this mean that the union relation considers "null matches null".  In other words, if I intersect `{NULL, 0, 5}` with `{10, NULL, 5}` is the result `{NULL, 5}` and if I intersect `{NULL, 0, 5}` with `{0, 10, 20}` is the result `{0}`?

Do we want to clarify that somewhere too (I think the "does NULL match NULL" question comes up in joins too but I don't remember if we clarified there or not)

**Comment by vbarua** [comment]
> Does this mean that the union relation considers "null matches null". 

@westonpace it does. I added a note about this.

**Review by EpsilonPrime — APPROVED** [review]

**Review by westonpace — APPROVED** [review]

**Pull request merged**